### PR TITLE
[JENKINS-49045] [Do not merge] Support whitelist of trusted fork PR authors

### PR DESCRIPTION
**Pull request by boonware**
Implements https://issues.jenkins-ci.org/browse/JENKINS-49045

### New Functionality
* See the above linked ticket for a description of the problem.
* Provided a single implementation of the `WhitelistSource` interface, the class `WhitelistGlobalConfiguration`. This provides a text area on the global Jenkins configuration page to define the whitelist. Further implementations could be added quite easily via the interface.
* Provided two new fork pull request trust mechanisms, "Whitelist" and "Contributors & Whitelist". These should be self-explanatory.

### What I Need Help With
* Dependency injection in Jenkins development is poorly documented. I've tried [this](https://wiki.jenkins.io/display/JENKINS/Dependency+Injection) and yet still get a NPE due to the unassigned dependencies. From my limited experience with Spring Boot I thought I could simply annotate the constructor with `@Inject`, and since there is only one `WhitelistSource` implementation available, the DI container (Guice?) would inject the dependency. I would appreciate any tips on how to get DI working in order to set the `whitelist` instance variable in my trust implementation classes assigned.

### Dependency Changes
* Introduced PowerMock to be able to mock the *final* method `GitHubSCMSourceRequest#getCollaboratorNames()` in unit tests.
* PowerMock required an upgrade to Mockito 2.X.

### Notes
* I notice the words "contributor" and "collaborator" are used interchangeably in this plugin. I believe the correct term should be "collaborator", i.e. someone who has write access to the repository. If this is confirmed as a defect I will create a new issue and update the code/strings in a separate PR.

### Screenshots of New UI Elements
1. New Trust Mechanisms
dropdown

2. Global Configuration
config

### CODE REVIEW

**Comment by boonware** [comment]
@stephenc @jglick Just @mentioning the top two committers to this repository. I think the feature here would provide a valuable enhancement of the plugin. I'm wracking my brains with DI in Jenkins however. Any pointers on how to get this working would be greatly appreciated.

File: pom.xml [hunk 1]
```diff
@@ -130,6 +131,18 @@
                 </exclusion>
             </exclusions>
         </dependency>
+        <dependency>
+            <groupId>org.powermock</groupId>
```
Comment (stephenc) [inline pom.xml]: please no, not powermock

File: src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java [hunk 2]
```diff
@@ -317,6 +319,111 @@ public boolean isApplicableToOrigin(@NonNull Class<? extends SCMHeadOrigin> orig
...
+     * An {@link SCMHeadAuthority} that trusts contributors to the repository as well as those on a whitelist.
+     */
+    public static class TrustContributorsAndWhitelist extends TrustContributors {
+
+        private WhitelistSource whitelist;
+
+        /**
+         * Constructor.
+         */
+        @Inject
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java]: do not use `@Inject`

File: src/main/java/org/jenkinsci/plugins/github_branch_source/WhitelistSource.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Set;
+
+/**
+ * Whitelists may be used for checking user permissions outside of the context of GitHub user
+ * permissions. Different whitelist sources may be defined, for example, a list of user IDs
+ * defined in a text field in the Jenkins UI.
+ *
+ * @since 2.3.2
+ */
+interface WhitelistSource {
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/WhitelistSource.java]: You will need this to be an abstract class that extends `AbstractDescribableImpl<WhitelistSourceDescriptor>` if i understand what you want to do

**Comment by stephenc** [comment]
Functionality belongs in an extension plugin not the core github branch source plugin

File: src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java (hunk 2 above)
Comment (boonware) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java]: @stephenc Would you mind explaining the correct way to do general purpose dependency injection in Jenkins? I'm writing this logic into a new plugin yet DI in Jenkins is poorly documented.

Comment (stephenc) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/ForkPullRequestDiscoveryTrait.java]: Don’t do dependency injection in Jenkins. Just do lookup by ExtensionList... much much safer

**Comment by boonware** [comment]
@stephenc Thanks for the tips. I was hoping you could give me a little help in solving the following problem:

I have a new implementation of `SCMHeadAuthority`. This implementation reads trusted PR authors from a whitelist. It will be possible to provide different whitelist sources, for example a text field in the global configuration page, an LDAP group etc. So at runtime the user will select which source they want on the global configuration page. The particular implementation will then need to be injected into the constructor of my `SCMHeadAuthority` instance. The steps will roughly be:

1, Lookup user's choice of whitelist source from configuration.
2. Create instance of that particular whitelist source.
3, How to pass this instance to the `SCMHeadAuthority` instance's constructor? The existing code in the GitHub Branch Source plugin is instantiating `SCMHeadAuthority` instances, none of which take constructor arguments, so I'm not sure if my approach will be compatible.

Of course, I could remove the constructor parameter and let the instance internally call a factory to get the correct whitelist instance, but obviously injecting the dependency will make things easier to test etc.

**Comment by stephenc** [comment]
Normally you instantiate via the UI binding which is `@DataBoundConstructor` (And sometimes `@DataBoundSetter` If you have optional config)

**Pull request closed without merging**